File: src/1387-Sort-Integers-by-The-Power-Value/1387.js
var getKth = function(lo, hi, k) {
    let res = [];
    for (let i = lo; i <= hi; i++) res[i - lo] = i;
    
    let cal = function(x) {
        let res = 0;
        while (x != 1) {
            if (x % 2 == 1) x = x * 3 + 1;
            else x >>= 1;
            res++;
        }
        return res;
    }
    
    res.sort((a, b) => {
        let p1 = cal(a), p2 = cal(b);
        return p1 == p2 ? a - b : p1 - p2;
    })
    return res[k - 1];
};

File: src/1387-Sort-Integers-by-The-Power-Value/1387.go
func getKth(lo int, hi int, k int) int {
    n := hi - lo + 1
    res := make([]int, n)
    for i := lo; i <= hi; i++ {
        res[i - lo] = i
    }
    
    sort.Slice(res, func(i, j int) bool {
        p1, p2 := cal(res[i]), cal(res[j])
        if p1 == p2 {
            return res[i] < res[j]
        } 
        return p1 < p2
    })
    return res[k - 1]
}

func cal(x int) int {
    res := 0
    for x != 1 {
        if x % 2 == 1 {
            x = x * 3 + 1
        } else {
            x >>= 1
        }
        res++
    }
    return res
}